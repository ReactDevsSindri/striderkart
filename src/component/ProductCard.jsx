import React, { useState } from "react";
import { NavLink } from "react-router-dom";
import { Button } from "../components/ui/button";

const ProductCard = ({ product }) => {
  console.log(product);
  const [img, setImg] = useState(product.images[0]);
  return (
    <div
      className=" product-list-thumbnail shadow dark:shadow-none dark:bg-slate-800 rounded-xl "
      style={{ position: "relative" }}
    >
      <NavLink to={`/productpage/${product._id}`} className="">
        <div
          className="group relative"
          onMouseLeave={() => setImg(product.images[0])}
        >
          <img src={img} alt="Poduct" />
          <div className="flex absolute bottom-0 opacity-0  group-hover:opacity-100 transition-opacity">
            <img
              src={product.images[0]}
              className="size-16 bg-slate-500"
              onMouseOver={() => setImg(product.images[0])}
            />
            <img
              src={product.images[1]}
              className="size-16 bg-slate-500"
              onMouseOver={() => setImg(product.images[1])}
            />
            <img
              src={product.images[2]}
              className="size-16 bg-slate-500"
              onMouseOver={() => setImg(product.images[2])}
            />
            <img
              src={product.images[3]}
              className="size-16 bg-slate-500"
              onMouseOver={() => setImg(product.images[3])}
            />
          </div>
        </div>
        <p
          className="m-0 mt-2 ps-3"
          style={{
            fontSize: "14px",
            fontWeight: "bold",
          }}
        >
          {product.productName?.substring(0, 25)}
        </p>
        <p className="m-0 mb-2 ps-3" style={{ fontSize: "12px" }}>
          2 colors
        </p>

        <Button style={{ position: "absolute", right: 10, bottom: 10 }}>
          Details
        </Button>
        {/* <img src={product.imgSrc} /> */}
      </NavLink>
    </div>
  );
};

export default ProductCard;
