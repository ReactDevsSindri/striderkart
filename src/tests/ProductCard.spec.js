import React from "react";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import ProductCard from "../component/ProductCard";
import renderer from "react-test-renderer";

describe("ProductCard component", () => {
  const product = {
    _id: "69b3258a-faf1-43af-a32a-41ad4e03b329",
    productName: "Shoe1",
    images: [
      "https://assets.adidas.com/images/h_840,f_auto,q_auto,fl_lossy,c_fill,g_auto/63c77c04dc6448548ccbae880189e107_9366/Galaxy_6_Shoes_Black_GW3848_01_standard.jpg",
      "https://assets.adidas.com/images/h_840,f_auto,q_auto,fl_lossy,c_fill,g_auto/63c77c04dc6448548ccbae880189e107_9366/Galaxy_6_Shoes_Black_GW3848_01_standard.jpg",
    ],
    productDesc: "Great shoe",
    sizes: [
      {
        size: "5",
        price: 599,
        _key: "17dd79cacd2a",
        offerPrice: 499,
      },
    ],
  };

  it("renders without crashing", () => {
    const productCard = render(
      <MemoryRouter>
        <ProductCard product={product} />
      </MemoryRouter>
    );
    expect(productCard).toMatchSnapshot();
  });

  it("changes the URL pathname when NavLink is clicked", () => {
    const { container } = render(
      <MemoryRouter>
        <ProductCard product={product} />
      </MemoryRouter>
    );

    const navLink = container.querySelector("a"); // Find the NavLink element

    fireEvent.click(navLink); // Simulate click on the NavLink
    setTimeout(() => {
      expect(window.location.pathname).toBe(
        `/productpage/${product.productId}`
      );
    }, 100);

    // Assert that the URL pathname has been updated
  });

  it('renders the "Details" button', () => {
    render(
      <MemoryRouter>
        <ProductCard product={product} />
      </MemoryRouter>
    );

    expect(screen.getByText("Details")).toBeInTheDocument();
  });

  it('performs expected action when "Details" button is clicked', async () => {
    const mockOnClick = jest.fn();
    const { container, history } = render(
      <MemoryRouter>
        <ProductCard product={product} onClick={mockOnClick} />
      </MemoryRouter>
    );

    fireEvent.click(screen.getByText("Details"));
    // Check that the correct path is navigated to
    setTimeout(() => {
      expect(history.location.pathname).toBe(
        `/productpage/${product.productId}`
      );
    }, 100);
  });
});
