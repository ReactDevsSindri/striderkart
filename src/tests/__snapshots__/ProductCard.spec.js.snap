// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard component renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class=" product-list-thumbnail shadow dark:shadow-none dark:bg-slate-800 rounded-xl "
        style="position: relative;"
      >
        <a
          class=""
          href="/productpage/69b3258a-faf1-43af-a32a-41ad4e03b329"
        >
          <div
            class="group relative"
          >
            <img
              alt="Poduct"
              src="https://assets.adidas.com/images/h_840,f_auto,q_auto,fl_lossy,c_fill,g_auto/63c77c04dc6448548ccbae880189e107_9366/Galaxy_6_Shoes_Black_GW3848_01_standard.jpg"
            />
            <div
              class="flex absolute bottom-0 opacity-0  group-hover:opacity-100 transition-opacity"
            >
              <img
                class="size-16 bg-slate-500"
                src="https://assets.adidas.com/images/h_840,f_auto,q_auto,fl_lossy,c_fill,g_auto/63c77c04dc6448548ccbae880189e107_9366/Galaxy_6_Shoes_Black_GW3848_01_standard.jpg"
              />
              <img
                class="size-16 bg-slate-500"
                src="https://assets.adidas.com/images/h_840,f_auto,q_auto,fl_lossy,c_fill,g_auto/63c77c04dc6448548ccbae880189e107_9366/Galaxy_6_Shoes_Black_GW3848_01_standard.jpg"
              />
              <img
                class="size-16 bg-slate-500"
              />
              <img
                class="size-16 bg-slate-500"
              />
            </div>
          </div>
          <p
            class="m-0 mt-2 ps-3"
            style="font-size: 14px; font-weight: bold;"
          >
            Shoe1
          </p>
          <p
            class="m-0 mb-2 ps-3"
            style="font-size: 12px;"
          >
            2 colors
          </p>
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
            style="position: absolute; right: 10px; bottom: 10px;"
          >
            Details
          </button>
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class=" product-list-thumbnail shadow dark:shadow-none dark:bg-slate-800 rounded-xl "
      style="position: relative;"
    >
      <a
        class=""
        href="/productpage/69b3258a-faf1-43af-a32a-41ad4e03b329"
      >
        <div
          class="group relative"
        >
          <img
            alt="Poduct"
            src="https://assets.adidas.com/images/h_840,f_auto,q_auto,fl_lossy,c_fill,g_auto/63c77c04dc6448548ccbae880189e107_9366/Galaxy_6_Shoes_Black_GW3848_01_standard.jpg"
          />
          <div
            class="flex absolute bottom-0 opacity-0  group-hover:opacity-100 transition-opacity"
          >
            <img
              class="size-16 bg-slate-500"
              src="https://assets.adidas.com/images/h_840,f_auto,q_auto,fl_lossy,c_fill,g_auto/63c77c04dc6448548ccbae880189e107_9366/Galaxy_6_Shoes_Black_GW3848_01_standard.jpg"
            />
            <img
              class="size-16 bg-slate-500"
              src="https://assets.adidas.com/images/h_840,f_auto,q_auto,fl_lossy,c_fill,g_auto/63c77c04dc6448548ccbae880189e107_9366/Galaxy_6_Shoes_Black_GW3848_01_standard.jpg"
            />
            <img
              class="size-16 bg-slate-500"
            />
            <img
              class="size-16 bg-slate-500"
            />
          </div>
        </div>
        <p
          class="m-0 mt-2 ps-3"
          style="font-size: 14px; font-weight: bold;"
        >
          Shoe1
        </p>
        <p
          class="m-0 mb-2 ps-3"
          style="font-size: 12px;"
        >
          2 colors
        </p>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
          style="position: absolute; right: 10px; bottom: 10px;"
        >
          Details
        </button>
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
